include/psa/crypto.h

```diff
@@ -4045,6 +4045,597 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ *
+ * \retval #PSA_SUCCESS
+ *         Operation completed successfully
+ *
+ * \retval #PSA_OPERATION_INCOMPLETE
+ *         Operation was interrupted due to the setting of \c
+ *         psa_interruptible_set_max_ops(), there is still work to be done,
+ *         please call this function again with the same operation object.
+ *
+ * \retval #PSA_ERROR_BUFFER_TOO_SMALL
+ *         The size of the \p signature buffer is too small. You can
+ *         determine a sufficient buffer size by calling
+ *         #PSA_SIGN_OUTPUT_SIZE(\c key_type, \c key_bits, \p alg)
+ *         where \c key_type and \c key_bits are the type and bit-size
+ *         respectively of \p key.
+ *
+ * \retval #PSA_ERROR_NOT_SUPPORTED
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ * \retval #PSA_ERROR_INSUFFICIENT_MEMORY
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
+ * \retval #PSA_ERROR_STORAGE_FAILURE
+ * \retval #PSA_ERROR_DATA_CORRUPT
+ * \retval #PSA_ERROR_DATA_INVALID
+ * \retval #PSA_ERROR_INSUFFICIENT_ENTROPY
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_sign_hash_complete(
+                             psa_sign_hash_interruptible_operation_t *operation,
+                             uint8_t *signature, size_t signature_size,
+                             size_t *signature_length );
+
+/** Abort a sign hash operation.
+ *
+ * Aborting an operation frees all associated resources except for the
+ * \p operation structure itself. Once aborted, the operation object
+ * can be reused for another operation by calling \c psa_sign_hash_start()
+ * again.
+ *
+ * You may call this function any time after the operation object has
+ * been initialized.
+ *
+ * In particular, calling \c psa_sign_hash_abort() after the operation has
+ * already been terminated by a call to \c psa_sign_hash_abort() or
+ * \c psa_sign_hash_complete() is safe and has no effect.
+ *
+ * \param[in,out] operation     Initialized sign hash operation.
+ *
+ * \retval #PSA_SUCCESS
+ *         The operation was aborted successfully.
+ *
+ * \retval #PSA_ERROR_NOT_SUPPORTED
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ * \retval #PSA_ERROR_INSUFFICIENT_MEMORY
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
+ * \retval #PSA_ERROR_STORAGE_FAILURE
+ * \retval #PSA_ERROR_DATA_CORRUPT
+ * \retval #PSA_ERROR_DATA_INVALID
+ * \retval #PSA_ERROR_INSUFFICIENT_ENTROPY
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_sign_hash_abort(
```

Comment:
Although the argument is sound, the result is uncompelling. We have `psa_sign_hash_start()` and `psa_sign_hash_complete()`, but then `psa_sign_hash_interruptible_abort()`. For consistency, we should insert 'interruptible' everywhere, but that also makes the API names more cumbersome than the currently succinct names for the multipart operations.

If we took a different approach, and used a different name for abandoning an interruptible operation - such as `psa_sign_hash_cancel()` - this might solve the multi-part sign-message vs interruptible sign-message concerns.

However, there are other legacy niggles that might complicate the naming issue, that is peculiar to signature algorithms:
* Of all the APIs, asymmetric signature is the only category that is greedy with the name-space. The other APIs are all of the form:
  -  `psa_<category>_<op>()` for the single part functions
  -  `psa_<category>_<sub-op>()` for multipart functions

  In contrast, we have `psa_<op>_<type>()` (where <op> is `sign` or `verify`) for asymmetric signature functions. In hindsight, we should probably have called these something like `psa_signature_sign_message()` etc., to be consistent.

*  Of all the multipart API functions, only the `psa_<category>_operation_init()` ones include the word 'operation'. Given the naming of all the other functions, we probably didn't need 'operation' in the name.

Another consideration is whether there are message signature algorithms that both:
1. Can be computed in a single pass over the message data, and
2. Require the key prior to processing the message data (for signature), or the key and/or signature prior to processing the message data (for verification)

At the moment, there is only a single algorithm in the Crypto API that fulfills both of these, and that is the PureEdDSA verify-message function. PureEdDSA sign-message requires two passes over the message data (which does not make it suitable for a generic multi-part sign-message function); and all other algorithms can be split into a hash operation and a sign-hash operation (that is RSA-PKCS#1-v1.5, RSA-PSS, ECDSA, HashEdDSA).

Is it worth building (or making namespace for) a multi-part message signature operation for PureEdDSA verification (and maybe other not-yet-implemented signature algorithms)? - given that there are reasonable application solutions to achieve this effect for other algorithms?